src/btree/bt_curnext.c

```diff
@@ -710,7 +688,12 @@ err:	switch (ret) {
 	case 0:
 		F_SET(cursor, WT_CURSTD_KEY_INT | WT_CURSTD_VALUE_INT);
 #ifdef HAVE_DIAGNOSTIC
-		ret = __wt_cursor_key_order_check(session, cbt, true);
+		/*
+		 * Skip key order check, if next is called after a prev returned
+		 * a prepare conflict error.
+		 */
+		if (!F_ISSET(cbt, WT_CBT_RETRY_PREV))
+			ret = __wt_cursor_key_order_check(session, cbt, true);
```

Comment:
I think we should clear `WT_CBT_RETRY_NEXT` and `WT_CBT_RETRY_PREV` in `__wt_btcur_iterate_setup()`, along with `WT_CBT_ITERATE_NEXT` and `WT_CBT_ITERATE_PREV`, in other words, if we're switching directions, we clear all of the iteration flags. (For the same reason, we might want to consider renaming `CBT_RETRY_NEXT/PREV` to be `CBT_ITERATE_RETRY_NEXT/PREV`, but I don't feel terribly strongly.)

Further, I think that means `__cursor_check_prepared_update()` should not be called as early in `__wt_btcur_next/prev`, that is, we'd optionally call `__wt_btcur_iterate_setup()`, then `__cursor_func_init()`, and then `__cursor_check_prepared_update`.

I'm hoping that change means this code is no longer necessary, but I'm not 100% sure of that. I really don't like this code appearing here, I'd like to see it go away, but I don't completely understand the need for it yet.